### PR TITLE
Fix UseLocalFunction when converting a single-line lambda.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/23872

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @Neme12

**Comment by CyrusNajmabadi** [comment]
Tagging @heejaechang since i changed some formatting code.  @heejaechang did i do this the right way?

Thanks!

**Comment by CyrusNajmabadi** [comment]
@Neme12 If @heejaechang likes how i did this, then you can merge this into your branch if you want.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide Merge plz :)

**Comment by Neme12** [comment]
isn't it a little excessive to have 9 test for a simple formatting change?

**Comment by CyrusNajmabadi** [comment]
@Neme12 I'm happy to add more tests.  Can you list what you'd like me to test here?

**Comment by CyrusNajmabadi** [comment]
> isn't it a little excessive to have 9 tests for a simple formatting change? 

I wanted to ensure that things would be preserved on a single line for simple/parenthesized lambdas.  And i wanted to make sure that none of the other code-gen parts of the feature would impact this.  i.e. formatting ties heavily into elastic trivia.  So i wanted to exercise a lot of the elastic trivia paths to ensure they would not break things.

**Comment by Neme12** [comment]
@CyrusNajmabadi no I think I added enough in my PR but now looking at what you added it seems to be excessive to focus on this simple change so much, there's already a lot of tests and I think they should each focus on a different feature, otherwise there would be thousands of them if we add tests for every possible combination of possible inputs

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi no I think I added enough in my PR but now looking at what you added it seems to be excessive to focus on this simple change so much

@Neme12 My goal when writing tests bugs is to both very that i've actually fixed the issue, and to ensure that regressions don't creep in in the future.  In this case, the feature has three very distinct paths it starts out with to generate the final local function.  i.e. "local and lambda on hte same line", "local and casted lambda on the same line" and "local in one statement, and initialized lambda in the next".  

Similarly, as i mentioned above, formatting is very finicky, and can be very affected by the presence of elastic trivia and the shape of your tree.

As such, it was important to me to test many permutations of these features working together.  It would not have surprised me at all to learn that my fix, for example, didn't work in the case where we needed to generate a parameter list from a simple lambda.  Or that my fix didn't work when we were merging two separate statements into one.  Or, even if my fix worked for those cases, we could easily regress in the future if the formatting engine changes.

> each new feature/fix would double the number of tests with this approach

We have a *ton* of experience testing Roslyn over the last decade (and the systems that predated it).  Devs on the team have a good understanding of the appropriate level of testing and when things get excessive, or when things are deemed to be testing things at the wrong level.  In this case i used my best judgement based on literally working with this codebase for tens of thousands of hours. :)  I stand by the decision here and i feel that the 9 tests are appropriate given the fix that was made, how the formatting subsystem works, and how the feature operates.

Thanks!

**Comment by CyrusNajmabadi** [comment]
@heejaechang Want to merge in?

**Comment by CyrusNajmabadi** [comment]
@heejaechang this doesn't actually work properly in several cases i've found.  The issue is that the formatting engine seems to want to ignore the "keep on single line" constraint if it finds any elastic trivia between the first two tokens of the result.  That's this code here: 

https://github.com/dotnet/roslyn/blob/b45885bb6c2fee5d1e996ee88ee88412dcc006a4/src/Workspaces/Core/Portable/Formatting/Context/FormattingContext.cs#L278-L282

There ends up being elastic trivia here because we are piecing together this local function from scratch, and we want to provide the starting nodes/tokens for it (i.e. generating the return type, or adding teh async modifiers).  Ensuring that there is no elastic trivia is non-trivial (no pun intended).  

What's the right way for a feature to say to the formatting engine: "i'm creating a local function.  please keep that on a single line if there are no actual newlines in it"?

**Comment by CyrusNajmabadi** [comment]
tagging @Neme12 as well as any suggestions from @heejaechang may be things we want to roll into his current PRs.

**Comment by heejaechang** [comment]
keep things in single line works if it sees it is on single line. when there is elastic trivia, it is not on single line. it is basically in unknown state. so keep single line won't work.

formatting engine itself won't do it for you . at least with default formatting rules. if feature wants its own behavior which is different than default behavior, it is either need to provide its own rule or manually modify trivia itself.

in this particular case, you can call node.NormalizeWhitespace() to remove all elastic trivia and ask formatting engine to format.

**Comment by CyrusNajmabadi** [comment]
>  it is either need to provide its own rule 

Ok.  What sort of rule would we provide here?

> node.NormalizeWhitespace()

NormalizeWhitespace will place a local function across multiple lines, somewhat defeating the purpose :)

**Comment by CyrusNajmabadi** [comment]
ok.  i think i have a working solution.

**Comment by heejaechang** [comment]
not sure where this came - SuppressOption.IgnoreElastic

but, looks like in default rule, some cases, we just have put that in. so LocalFunction might be one of those case as well?

File: src/Workspaces/CSharp/Portable/Formatting/Rules/SuppressFormattingRule.cs [hunk 1]
```diff
@@ -124,9 +124,11 @@ private void AddSpecificNodesSuppressOperations(List<SuppressOperation> list, Sy
                 }
             }
 
-            if (node is AnonymousMethodExpressionSyntax anonymousMethod)
+            if (node is AnonymousFunctionExpressionSyntax ||
+                node is LocalFunctionStatementSyntax)
             {
-                AddSuppressWrappingIfOnSingleLineOperation(list, anonymousMethod.DelegateKeyword, anonymousMethod.GetLastToken(includeZeroWidth: true));
+                AddSuppressWrappingIfOnSingleLineOperation(
```
Comment (heejaechang) [inline src/Workspaces/CSharp/Portable/Formatting/Rules/SuppressFormattingRule.cs]: you can put that IgnoreElastic option here. I am not sure why default rule would make some of construct to ignore elastic trivia, but looks like we already have some of those. so probably fine.

Comment (heejaechang) [reply src/Workspaces/CSharp/Portable/Formatting/Rules/SuppressFormattingRule.cs]: http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/Formatting/Rules/ElasticTriviaFormattingRule.cs,46

Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Formatting/Rules/SuppressFormattingRule.cs]: What about https://github.com/dotnet/roslyn/pull/24867 as the potential solution?

**Comment by heejaechang** [comment]
> Ok. What sort of rule would we provide here?

you can call Formatter.Format(node, new MyOwnRule().Concat(Formatter.DefaultRules...)

and myOwnRule having 
https://github.com/dotnet/roslyn/pull/24773/files/cf67fec1716b6818bde5deb8da82f1c49a447b9e#diff-17d57a33edf1c6d37a37ac7e27c947a2
with IgnoreElastic option, but again, I think by looking existing rule, I think you can just add IgnoreElastic option in the default rule. and done with ti.

looks like we already do that for property and anonymous function and etc. so feels like it is not adding something different than what we already have.

File: src/Workspaces/CSharp/Portable/Formatting/Rules/SuppressFormattingRule.cs (hunk 1 above)
Comment (heejaechang) [reply src/Workspaces/CSharp/Portable/Formatting/Rules/SuppressFormattingRule.cs]: that will work but, have you tried just adding IgnoreElastic option to the rule you added? that feels more simple. and looks like 
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Formatting/Rules/Operations/SuppressOption.cs,2a0df9bb73a427bd,references

we already have such rules.

Comment (heejaechang) [reply src/Workspaces/CSharp/Portable/Formatting/Rules/SuppressFormattingRule.cs]: okay, let me revise my answer. if you think this local single line case is specific to this code fix, then +1 on what you did (removing all elastic trivia)

if you think, it is better behavior in general, I think you should add that case in ElasticTriviaFormattingRule with IgnoreElastic option.

**Comment by CyrusNajmabadi** [comment]
that makes sense.  @Neme12 do you want to try that?

File: src/Workspaces/CSharp/Portable/Formatting/Rules/SuppressFormattingRule.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Formatting/Rules/SuppressFormattingRule.cs]: Yeah, i think the behavior is better in general.  So i went with teh approach you had for properties.

**Comment by Neme12** [comment]
@CyrusNajmabadi it looks like you already did exactly that in your new PR. should i wait for it to get merged?

**Pull request merged**